### PR TITLE
Support IDENTITY() function in SELECT-INTO

### PR DESCRIPTION
**Pull request by deepakshi-mittal**
Changes for Identity in Select Into statement
Added hooks for identity
Added fields in intoclause to pass information about identity column

Engine Pr: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/166

Task : BABEL-539

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 1]
```diff
@@ -3391,6 +3391,9 @@ CREATE OR REPLACE FUNCTION sys.host_id()
 RETURNS sys.VARCHAR(10)  AS 'babelfishpg_tsql' LANGUAGE C IMMUTABLE PARALLEL SAFE;
 GRANT EXECUTE ON FUNCTION sys.host_id() TO PUBLIC;
 
+CREATE OR REPLACE FUNCTION sys.identity_into(IN typename INT, IN seed INT, IN increment INT)
+RETURNS int AS 'babelfishpg_tsql' LANGUAGE C STABLE;
+
```
Comment (macwoody) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:3397]: Here is a dilemma you are facing. 
If you don't grant permission to public, non-dbo users would not be able to call this function. Thus could not use your feature. 
If you grant, everyone would be able to call, which would move a sequence forward bypassing security check.

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c [hunk 2]
```diff
@@ -230,6 +230,36 @@ TsqlFunctionConvert(TypeName *typename, Node *arg, Node *style, bool try, int lo
...
+	Oid base_oid;
+	typenameTypeIdAndMod(NULL, typename, &type_oid, &typmod);
+	base_oid =  getBaseType(type_oid);
+	switch (base_oid)
+		{
+			case INT2OID:
+			case INT4OID:
+			case INT8OID:
+			case NUMERICOID:				
+				args = list_make3((Node *)makeIntConst((int)type_oid, -1), seed, increment);
```
Comment (macwoody) [inline contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c]: why -1?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 3]
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
...
+			type_string = "pg_catalog.int4";
+			break;
+		case INT8OID:
+			type_string = "pg_catalog.int8";
+			break;
+		case NUMERICOID:
+			type_string = "pg_catalog.numeric";
+			break;
+		default:
+			type_string = "decimal";
```
Comment (macwoody) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: why do we have decimal as default? Shouldn't it be an error here?

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp [hunk 4]
```diff
@@ -1846,6 +1847,17 @@ class tsqlBuilder : public tsqlCommonMutator
 		}
 	}
 
+	void enterFunction_call(TSqlParser::Function_callContext *ctx) override
+	{
+		std::string func_name = ::getFullText(ctx);
```
Comment (macwoody) [inline contrib/babelfishpg_tsql/src/tsqlIface.cpp]: I think you don't have to do this way. 
There is more elegant way of getting function name. 
You could find helper functions to do so.

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 5]
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
...
+			break;
+	}
+	return type_string;
+}
+
+static List * transformSelectIntoStmt(CreateTableAsStmt *stmt, const char *queryString){
```
Comment (macwoody) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Please try drop dependent sequence explicitly and see if the dropping sequence could be blocked. 
And also check if the sequence is gone after dropping the table.

**Review by macwoody — DISMISSED** [review]

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 5 above)
Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Yes when we try to drop sequence explicitly it errors out saying that can't drop a dependent sequence.
Yes sequence is gone after we drop the table.

File: contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c (hunk 2 above)
Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/src/backend_parser/gram-tsql-epilogue.y.c]: its for location, I will change it

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 1 above)
Comment (forestkeeper) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:3397]: could be 
1. grant to public , and then 
2. add a block in parser to block customer from calling this func directly

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 3 above)
Comment (forestkeeper) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: +1
Should throw error for other types, and add a case for decimal

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 6]
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
...
+
+	if (n && n->type == T_Query)
+	{
+		Query *q = (Query *) n;
+		bool seen_identity = false;
+		foreach(elements, q->targetList)
+		{	
+			TargetEntry *tle = (TargetEntry *) lfirst(elements);
+			FuncExpr *funcexpr;
+			if (tle->expr && IsA(tle->expr, FuncExpr) && strcasecmp(get_func_name(((FuncExpr *) (tle->expr))->funcid), "identity_into") ==0 ){
```
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: we should write 
```
if ()
{
}
```

not 
```
if(){
}
```
please make { not the same line with if ()

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 7]
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
...
+							break;
+						case 2:
+							con = (Const *) fargNode;
+							seqoptions = lappend(seqoptions, makeDefElem("start", (Node *)makeInteger(con->constvalue), -1));
+							seed_value = (int64) con->constvalue;
+							break;
+						case 3:
+							con = (Const *) fargNode;
+							seqoptions = lappend(seqoptions,  makeDefElem("increment", (Node *)makeInteger(con->constvalue), -1));
+							if ((int) con->constvalue > 0){
```
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Should update the if else format to 
```
if()
{
}
else
{
}
```

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 8]
```diff
@@ -5022,6 +5026,191 @@ pltsql_revert_guc(int nest_level)
...
+void
+pltsql_bbfSelectIntoUtility(ParseState *pstate, PlannedStmt *pstmt, const char *queryString, QueryEnvironment *queryEnv, 
+	ParamListInfo params, QueryCompletion *qc){
+
+	Node *parsetree = pstmt->utilityStmt;	
+	ObjectAddress address;
+	ObjectAddress secondaryObject = InvalidObjectAddress;
+	List *stmts;
+	stmts = transformSelectIntoStmt((CreateTableAsStmt *) parsetree, queryString);
+		while (stmts != NIL)
```
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: the tab interval is not right from this line

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 1 above)
Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:3397]: Done 
grant to public , and then
add a block in parser to block customer from calling this func directly

File: contrib/babelfishpg_tsql/src/tsqlIface.cpp (hunk 4 above)
Comment (deepakshi-mittal) [reply contrib/babelfishpg_tsql/src/tsqlIface.cpp]: Done used helper functions

**Review by forestkeeper — APPROVED** [review]
lgtm

**Pull request merged**